File: SlimFileManager/src/main/java/com/slim/slimfilemanager/utils/FileUtil.java
package com.slim.slimfilemanager.utils;

import android.content.Context;
import android.net.Uri;
import android.os.Build;
import android.os.ParcelFileDescriptor;
import android.support.v4.provider.DocumentFile;
import android.telephony.IccOpenLogicalChannelResponse;
import android.text.TextUtils;

import com.slim.slimfilemanager.settings.SettingsProvider;

import org.apache.commons.io.FileUtils;
import org.apache.commons.io.FilenameUtils;
import org.apache.commons.io.IOUtils;

import java.io.DataOutputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStream;
import java.nio.charset.Charset;

import trikita.log.Log;

public class FileUtil {

    public static boolean copyFile(Context context, String f, String fol) {
        File file = new File(f);
        File folder = new File(fol);

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            FileInputStream fis = null;
            OutputStream os = null;
            try {
                fis = new FileInputStream(file);
                DocumentFile target = DocumentFile.fromFile(folder);
                os = context.getContentResolver().openOutputStream(target.getUri());
                IOUtils.copy(fis, os);
                return true;
            } catch (Exception e) {
                e.printStackTrace();
            } finally {
                try {
                    if (fis != null) {
                        fis.close();
                    }
                    if (os != null) {
                        os.flush();
                        os.close();
                    }
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        }

        if (file.exists()) {
            if (!folder.exists()) {
                if (!mkdir(context, folder)) return false;
            }
            try {
                if (file.isDirectory()) {
                    FileUtils.copyDirectoryToDirectory(file, folder);
                } else if (file.isFile()) {
                    FileUtils.copyFileToDirectory(file, folder);
                }
                return true;
            } catch (IOException e) {
                return SettingsProvider.getBoolean(context, SettingsProvider.KEY_ENABLE_ROOT, false)
                        && RootUtils.isRootAvailable() && RootUtils.copyFile(f, fol);
            }
        } else {
            return false;
        }
    }

    public static boolean moveFile(Context context, String source, String destination) {
        if (TextUtils.isEmpty(source) || TextUtils.isEmpty(destination)) {
            return false;
        }
        File file = new File(source);
        File folder = new File(destination);

        try {
            FileUtils.moveFileToDirectory(file, folder, true);
            return true;
        } catch (IOException e) {
            return SettingsProvider.getBoolean(context, SettingsProvider.KEY_ENABLE_ROOT, false)
                    && RootUtils.isRootAvailable() && RootUtils.moveFile(source, destination);
        }
    }

    public static boolean deleteFile(Context context, String path) {
        try {
            FileUtils.forceDelete(new File(path));
            return true;
        } catch (IOException e) {
            return SettingsProvider.getBoolean(context, SettingsProvider.KEY_ENABLE_ROOT, false)
                    && RootUtils.isRootAvailable() && RootUtils.deleteFile(path);
        }
    }

    public static String[] getFileProperties(Context context, File file) {
        String[] info = null;

        RootUtils.CommandOutput out;
        if (SettingsProvider.getBoolean(context, SettingsProvider.KEY_ENABLE_ROOT, false)
                && RootUtils.isRootAvailable()) {
            out = RootUtils.runCommand("ls -l " + file.getAbsolutePath());
        } else {
            out = runCommand("ls -l " + file.getAbsolutePath());
        }
        if (out == null) return null;
        if (TextUtils.isEmpty(out.error) && out.exitCode == 0) {
            info = getAttrs(out.output);
        }
        return info;
    }

    private static String[] getAttrs(String string) {
        if (string.length() < 44) {
            throw new IllegalArgumentException("Bad ls -l output: " + string);
        }
        final char[] chars = string.toCharArray();

        final String[] results = new String[11];
        int ind = 0;
        final StringBuilder current = new StringBuilder();

        Loop:
        for (int i = 0; i < chars.length; i++) {
            switch (chars[i]) {
                case ' ':
                case '\t':
                    if (current.length() != 0) {
                        results[ind] = current.toString();
                        ind++;
                        current.setLength(0);
                        if (ind == 10) {
                            results[ind] = string.substring(i).trim();
                            break Loop;
                        }
                    }
                    break;

                default:
                    current.append(chars[i]);
                    break;
            }
        }

        return results;
    }

    public static RootUtils.CommandOutput runCommand(String cmd) {
        RootUtils.CommandOutput output = new RootUtils.CommandOutput();
        try {
            Process process = Runtime.getRuntime().exec("sh");
            DataOutputStream os = new DataOutputStream(
                    process.getOutputStream());
            os.writeBytes(cmd + "\n");
            os.writeBytes("exit\n");
            os.flush();

            output.exitCode = process.waitFor();
            output.output = IOUtils.toString(process.getInputStream());
            output.error = IOUtils.toString(process.getErrorStream());

            if (output.exitCode != 0 || (!TextUtils.isEmpty(output.error))) {
                Log.e("Root Error, cmd: " + cmd, "error: " + output.error);
            }
        } catch (IOException | InterruptedException e) {
            e.printStackTrace();
        }
        return output;
    }

    public static boolean changeGroupOwner(Context context, File file, String owner, String group) {
        try {
            boolean useRoot = false;
            if (!file.canWrite() && SettingsProvider.getBoolean(context,
                    SettingsProvider.KEY_ENABLE_ROOT, false)
                    && RootUtils.isRootAvailable()) {
                useRoot = true;
                RootUtils.remountSystem("rw");
            }

            if (useRoot) {
                RootUtils.runCommand("chown " + owner + "." + group + " "
                        + file.getAbsolutePath());
            } else {
                runCommand("chown " + owner + "." + group + " "
                        + file.getAbsolutePath());
            }
            return true;
        } catch (Exception e) {
            e.printStackTrace();
        }

        return false;
    }

    public static boolean applyPermissions(Context context, File file, Permissions permissions) {
        try {
            boolean useSu = false;
            if (!file.canWrite() && SettingsProvider.getBoolean(context,
                    SettingsProvider.KEY_ENABLE_ROOT, false)
                    && RootUtils.isRootAvailable()) {
                useSu = true;
                RootUtils.remountSystem("rw");
            }

            if (useSu) {
                RootUtils.runCommand("chmod " + permissions.getOctalPermissions() + " "
                        + file.getAbsolutePath());
            } else {
                runCommand("chmod " + permissions.getOctalPermissions() + " "
                        + file.getAbsolutePath());
            }
            return true;
        } catch (Exception e) {
            e.printStackTrace();
        }

        return false;
    }

    public static boolean mkdir(Context context, File dir) {
        if (dir.exists()) {
            return false;
        }
        if (dir.mkdirs()) {
            return true;
        } else {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                DocumentFile document = DocumentFile.fromFile(dir.getParentFile());
                if (document.exists()) {
                    if (document.createDirectory(dir.getAbsolutePath()).exists()) {
                        return true;
                    }
                }
            }
        }

        return (SettingsProvider.getBoolean(context,
                SettingsProvider.KEY_ENABLE_ROOT, false) && RootUtils.isRootAvailable()
                && RootUtils.createFolder(dir));
    }

    public static String getExtension(String fileName) {
        return FilenameUtils.getExtension(fileName);
    }

    public static String removeExtension(String s) {
        return FilenameUtils.removeExtension(new File(s).getName());
    }

    public static void writeFile(Context context, String content, File file, String encoding) {
        if (file.canWrite()) {
            try {
                FileUtils.write(file, content, encoding);
            } catch (IOException e) {
                // ignore
            }
        } else if (SettingsProvider.getBoolean(context, SettingsProvider.KEY_ENABLE_ROOT, false)) {
            RootUtils.writeFile(file, content);
        }
    }

    public static void renameFile(Context context, File oldFile, File newFile) {
        if (oldFile.renameTo(newFile)) {
            return;
        } else {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                DocumentFile document = DocumentFile.fromFile(oldFile);
                if (document.renameTo(newFile.getAbsolutePath())) {
                    return;
                }
            }
        }
        if (SettingsProvider.getBoolean(context, SettingsProvider.KEY_ENABLE_ROOT, false)) {
            RootUtils.renameFile(oldFile, newFile);
        }
    }

    public static long getFileSize(File file) {
        return file.length();
    }

    public static void writeUri(Context context,
                                Uri uri, String newContent, String encoding) throws IOException {
        ParcelFileDescriptor pfd = context.getContentResolver().openFileDescriptor(uri, "w");
        if (pfd != null) {
            FileOutputStream fileOutputStream = new FileOutputStream(pfd.getFileDescriptor());
            fileOutputStream.write(newContent.getBytes(Charset.forName(encoding)));
            fileOutputStream.close();
            pfd.close();
        }
    }
}

File: manager/src/main/java/com/slim/slimfilemanager/utils/FileUtil.kt
package com.slim.slimfilemanager.utils

import android.content.Context
import android.os.Build
import android.support.v4.provider.DocumentFile
import android.text.TextUtils
import android.util.Log

import com.slim.slimfilemanager.settings.SettingsProvider

import org.apache.commons.io.FileUtils
import org.apache.commons.io.FilenameUtils
import org.apache.commons.io.IOUtils

import java.io.DataOutputStream
import java.io.File
import java.io.FileInputStream
import java.io.IOException
import java.io.OutputStream

object FileUtil {

    fun copyFile(context: Context, f: String, fol: String): Boolean {
        val file = File(f)
        val folder = File(fol)

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            var fis: FileInputStream? = null
            var os: OutputStream? = null
            try {
                fis = FileInputStream(file)
                val target = DocumentFile.fromFile(folder)
                os = context.contentResolver.openOutputStream(target.uri)
                IOUtils.copy(fis, os)
                return true
            } catch (e: Exception) {
                e.printStackTrace()
            } finally {
                try {
                    fis?.close()
                    if (os != null) {
                        os.flush()
                        os.close()
                    }
                } catch (e: IOException) {
                    e.printStackTrace()
                }

            }
        }

        if (file.exists()) {
            if (!folder.exists()) {
                if (!mkdir(context, folder)) return false
            }
            try {
                if (file.isDirectory) {
                    FileUtils.copyDirectoryToDirectory(file, folder)
                } else if (file.isFile) {
                    FileUtils.copyFileToDirectory(file, folder)
                }
                return true
            } catch (e: IOException) {
                return (SettingsProvider.getBoolean(context, SettingsProvider.KEY_ENABLE_ROOT,
                        false)
                        && RootUtils.isRootAvailable && RootUtils.copyFile(f, fol))
            }

        } else {
            return false
        }
    }

    fun moveFile(context: Context, source: String, destination: String): Boolean {
        if (TextUtils.isEmpty(source) || TextUtils.isEmpty(destination)) {
            return false
        }
        val file = File(source)
        val folder = File(destination)

        try {
            FileUtils.moveFileToDirectory(file, folder, true)
            return true
        } catch (e: IOException) {
            return (SettingsProvider.getBoolean(context, SettingsProvider.KEY_ENABLE_ROOT, false)
                    && RootUtils.isRootAvailable && RootUtils.moveFile(source, destination))
        }

    }

    fun deleteFile(context: Context, path: String): Boolean {
        try {
            FileUtils.forceDelete(File(path))
            return true
        } catch (e: IOException) {
            return (SettingsProvider.getBoolean(context, SettingsProvider.KEY_ENABLE_ROOT, false)
                    && RootUtils.isRootAvailable && RootUtils.deleteFile(path))
        }

    }

    fun getFileProperties(context: Context, file: File): Array<String>? {
        var info: Array<String>? = null

        val out: RootUtils.CommandOutput?
        if (SettingsProvider.getBoolean(context, SettingsProvider.KEY_ENABLE_ROOT,
                        false) && RootUtils.isRootAvailable) {
            out = RootUtils.runCommand("ls -l " + file.absolutePath)
        } else {
            out = runCommand("ls -l " + file.absolutePath)
        }
        if (out == null) return null
        if (TextUtils.isEmpty(out.error) && out.exitCode == 0) {
            info = getAttrs(out.output!!)
        }
        return info
    }

    private fun getAttrs(string: String): Array<String> {
        if (string.length < 44) {
            throw IllegalArgumentException("Bad ls -l output: $string")
        }
        val chars = string.toCharArray()

        val results = emptyArray<String>()
        var ind = 0
        val current = StringBuilder()

        Loop@ for (i in chars.indices) {
            when (chars[i]) {
                ' ', '\t' -> if (current.length != 0) {
                    results[ind] = current.toString()
                    ind++
                    current.setLength(0)
                    if (ind == 10) {
                        results[ind] = string.substring(i).trim({ it <= ' ' })
                        break@Loop
                    }
                }

                else -> current.append(chars[i])
            }
        }

        return results
    }

    fun runCommand(cmd: String): RootUtils.CommandOutput {
        val output = RootUtils.CommandOutput()
        try {
            val process = Runtime.getRuntime().exec("sh")
            val os = DataOutputStream(
                    process.outputStream)
            os.writeBytes(cmd + "\n")
            os.writeBytes("exit\n")
            os.flush()

            output.exitCode = process.waitFor()
            output.output = IOUtils.toString(process.inputStream)
            output.error = IOUtils.toString(process.errorStream)

            if (output.exitCode != 0 || !TextUtils.isEmpty(output.error)) {
                Log.e("Shell Error, cmd: $cmd", "error: " + output.error!!)
            }
        } catch (e: IOException) {
            e.printStackTrace()
        } catch (e: InterruptedException) {
            e.printStackTrace()
        }

        return output
    }

    fun changeGroupOwner(context: Context, file: File, owner: String, group: String): Boolean {
        try {
            var useRoot = false
            if (!file.canWrite() && SettingsProvider.getBoolean(context,
                            SettingsProvider.KEY_ENABLE_ROOT, false)
                    && RootUtils.isRootAvailable) {
                useRoot = true
                RootUtils.remountSystem("rw")
            }

            if (useRoot) {
                RootUtils.runCommand("chown " + owner + "." + group + " "
                        + file.absolutePath)
            } else {
                runCommand("chown " + owner + "." + group + " "
                        + file.absolutePath)
            }
            return true
        } catch (e: Exception) {
            e.printStackTrace()
        }

        return false
    }

    fun applyPermissions(context: Context, file: File, permissions: Permissions): Boolean {
        try {
            var useSu = false
            if (!file.canWrite() && SettingsProvider.getBoolean(context,
                            SettingsProvider.KEY_ENABLE_ROOT, false)
                    && RootUtils.isRootAvailable) {
                useSu = true
                RootUtils.remountSystem("rw")
            }

            if (useSu) {
                RootUtils.runCommand("chmod " + permissions.octalPermissions + " "
                        + file.absolutePath)
            } else {
                runCommand("chmod " + permissions.octalPermissions + " "
                        + file.absolutePath)
            }
            return true
        } catch (e: Exception) {
            e.printStackTrace()
        }

        return false
    }

    fun mkdir(context: Context, dir: File): Boolean {
        if (dir.exists()) {
            return false
        }
        if (dir.mkdirs()) {
            return true
        } else {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                val document = DocumentFile.fromFile(dir.parentFile)
                if (document.exists()) {
                    if (document.createDirectory(dir.absolutePath)!!.exists()) {
                        return true
                    }
                }
            }
        }

        return (SettingsProvider.getBoolean(context,
                SettingsProvider.KEY_ENABLE_ROOT, false) && RootUtils.isRootAvailable
                && RootUtils.createFolder(dir))
    }

    fun getExtension(fileName: String): String {
        return FilenameUtils.getExtension(fileName)
    }

    fun removeExtension(s: String): String {
        return FilenameUtils.removeExtension(File(s).name)
    }

    fun writeFile(context: Context, content: String, file: File, encoding: String) {
        if (file.canWrite()) {
            try {
                FileUtils.write(file, content, encoding)
            } catch (e: IOException) {
                // ignore
            }

        } else if (SettingsProvider.getBoolean(context, SettingsProvider.KEY_ENABLE_ROOT, false)) {
            RootUtils.writeFile(file, content)
        }
    }

    fun renameFile(context: Context, oldFile: File, newFile: File) {
        if (oldFile.renameTo(newFile)) {
            return
        } else {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                val document = DocumentFile.fromFile(oldFile)
                if (document.renameTo(newFile.absolutePath)) {
                    return
                }
            }
        }
        if (SettingsProvider.getBoolean(context, SettingsProvider.KEY_ENABLE_ROOT, false)) {
            RootUtils.renameFile(oldFile, newFile)
        }
    }

    fun getFileSize(file: File): Long {
        return file.length()
    }
}
